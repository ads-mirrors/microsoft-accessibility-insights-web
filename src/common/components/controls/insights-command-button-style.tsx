// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { makeStyles, tokens } from '@fluentui/react-components';

export const useInsightsCommandButtonStyle = makeStyles({
    button: {
        fontWeight: tokens?.fontWeightRegular,
    },

    menuItem: {
        border: 'unset !important',
    },
});
