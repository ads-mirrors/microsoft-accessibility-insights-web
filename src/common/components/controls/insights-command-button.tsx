// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { mergeClasses, MenuButtonProps, Button, MenuItem } from '@fluentui/react-components';

import { useInsightsCommandButtonStyle } from 'common/components/controls/insights-command-button-style';
import { NamedFC } from 'common/react/named-fc';
import * as React from 'react';
import styles from './insights-command-button.scss';

export type InsightsCommandButtonIconProps = {
    className?: string;
    icon?: JSX.Element;
};

export type InsightsCommandButtonProps = (MenuButtonProps | any) & {
    insightsCommandButtonIconProps?: InsightsCommandButtonIconProps;
    isNarrowMode?: boolean;
};

// See https://www.figma.com/file/Wj4Ggf6GGQBQkiDIaHfXRX2B/Accessibility-Insights%3A-Styles?node-id=1%3A27
export const InsightsCommandButton = NamedFC<InsightsCommandButtonProps>(
    'InsightsCommandButton',
    React.forwardRef((props, ref) => {
        const overrides = useInsightsCommandButtonStyle();
        return props.isNarrowMode ? (
            <MenuItem
                className={overrides.menuItem}
                icon={{
                    children: props.insightsCommandButtonIconProps?.icon,
                }}
                ref={ref}
                {...props}
            >
                {props.children}
            </MenuItem>
        ) : (
            <Button
                appearance="transparent"
                className={mergeClasses(
                    styles?.insightsCommandButton,
                    props.className,
                    overrides?.button,
                )}
                shape="square"
                icon={{
                    children: props.insightsCommandButtonIconProps?.icon,
                }}
                ref={ref}
                {...props}
            >
                {props.children}
            </Button>
        );
    }),
);
