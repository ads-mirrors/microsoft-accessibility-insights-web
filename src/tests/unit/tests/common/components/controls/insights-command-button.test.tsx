// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Button } from '@fluentui/react-components';
import { FolderOpenRegular } from '@fluentui/react-icons';

import { render } from '@testing-library/react';
import { InsightsCommandButton } from 'common/components/controls/insights-command-button';
import * as React from 'react';
import {
    expectMockedComponentPropsToMatchSnapshots,
    mockReactComponents,
} from '../../../../mock-helpers/mock-module-helpers';

jest.mock('common/components/controls/insights-command-button-style', () => {
    return {
        useInsightsCommandButtonStyle: jest.fn(),
    };
});

jest.mock('@fluentui/react-components');

describe('InsightsCommandButton', () => {
    mockReactComponents([Button]);
    const props = {
        insightsCommandButtonIconProps: {
            className: 'startOverMenuItemIcon',
        },
        className: '',
    };
    it('renders per snapshot with props passed through', () => {
        const renderResult = render(
            <InsightsCommandButton {...props}>test-text</InsightsCommandButton>,
        );
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([Button]);
    });

    it('renders per snapshot with extra className combined with its own', () => {
        const renderResult = render(<InsightsCommandButton {...props} />);
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([Button]);
    });

    it('renders per snapshot with iconProps passed through', () => {
        const renderResult = render(
            <InsightsCommandButton insightsCommandButtonIconProps={{ icon: <FolderOpenRegular /> }}>
                Load assessment
            </InsightsCommandButton>,
        );
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([Button]);
    });

    it('renders per snapshot with extra icon className combined with its own', () => {
        const renderResult = render(
            <InsightsCommandButton
                insightsCommandButtonIconProps={{
                    icon: <FolderOpenRegular className="icon-class-name" />,
                }}
            >
                Load assessment
            </InsightsCommandButton>,
        );
        expect(renderResult.asFragment()).toMatchSnapshot();
        expectMockedComponentPropsToMatchSnapshots([Button]);
    });
});
