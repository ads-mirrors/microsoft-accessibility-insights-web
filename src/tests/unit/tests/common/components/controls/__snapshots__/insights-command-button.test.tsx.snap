// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InsightsCommandButton renders per snapshot with extra className combined with its own 1`] = `
<DocumentFragment>
  <mock-button
    appearance="transparent"
    classname=""
    icon="[object Object]"
    insightscommandbuttoniconprops="[object Object]"
    shape="square"
  />
</DocumentFragment>
`;

exports[`InsightsCommandButton renders per snapshot with extra className combined with its own: Button props 1`] = `
{
  "appearance": "transparent",
  "children": undefined,
  "className": "",
  "icon": {
    "children": undefined,
  },
  "insightsCommandButtonIconProps": {
    "className": "startOverMenuItemIcon",
  },
  "shape": "square",
}
`;

exports[`InsightsCommandButton renders per snapshot with extra icon className combined with its own 1`] = `
<DocumentFragment>
  <mock-button
    appearance="transparent"
    icon="[object Object]"
    insightscommandbuttoniconprops="[object Object]"
    shape="square"
  >
    Load assessment
  </mock-button>
</DocumentFragment>
`;

exports[`InsightsCommandButton renders per snapshot with extra icon className combined with its own: Button props 1`] = `
{
  "appearance": "transparent",
  "children": "Load assessment",
  "className": undefined,
  "icon": {
    "children": <FolderOpenRegular
      className="icon-class-name"
    />,
  },
  "insightsCommandButtonIconProps": {
    "icon": <FolderOpenRegular
      className="icon-class-name"
    />,
  },
  "shape": "square",
}
`;

exports[`InsightsCommandButton renders per snapshot with iconProps passed through 1`] = `
<DocumentFragment>
  <mock-button
    appearance="transparent"
    icon="[object Object]"
    insightscommandbuttoniconprops="[object Object]"
    shape="square"
  >
    Load assessment
  </mock-button>
</DocumentFragment>
`;

exports[`InsightsCommandButton renders per snapshot with iconProps passed through: Button props 1`] = `
{
  "appearance": "transparent",
  "children": "Load assessment",
  "className": undefined,
  "icon": {
    "children": <FolderOpenRegular />,
  },
  "insightsCommandButtonIconProps": {
    "icon": <FolderOpenRegular />,
  },
  "shape": "square",
}
`;

exports[`InsightsCommandButton renders per snapshot with props passed through 1`] = `
<DocumentFragment>
  <mock-button
    appearance="transparent"
    classname=""
    icon="[object Object]"
    insightscommandbuttoniconprops="[object Object]"
    shape="square"
  >
    test-text
  </mock-button>
</DocumentFragment>
`;

exports[`InsightsCommandButton renders per snapshot with props passed through: Button props 1`] = `
{
  "appearance": "transparent",
  "children": "test-text",
  "className": "",
  "icon": {
    "children": undefined,
  },
  "insightsCommandButtonIconProps": {
    "className": "startOverMenuItemIcon",
  },
  "shape": "square",
}
`;
